### ISSUE TITLE
Subclassing Solution vs. using the attributes map.

### ISSUE THREAD

**Issue by fstonedahl** [MOEAFramework/MOEAFramework#125]
A student and I are interested in extending MOEA to include Novelty Search (http://eplex.cs.ucf.edu/noveltysearch/userspage/), which evaluates fitness by comparing some sort of phenotypic behavior to the behaviors that have been observed before.  Rather curiously, it avoids using the objectives in the fitness computation at all.  Instead, individuals are considered highly "fit" if they show novel behavior (different from previously archived behaviors).  To implement this, we'll need to store behavior data (probably a real-valued vector) in each solution.

I see two basic approaches:

1. We could subclass Solution, and add a double[] behavior field.
2. We could store the double[] using a "BEHAVIOR" key in the attributes Map.

I feel like **behavior** ought to be treated the same as the **objectives** field, and I see that temporary attributes  are *not* copied over when solutions are cloned.  However, I don't have a good sense of whether this actually matters.  i.e. *Why* is it important for the _objectives_ to be copied over?  Aren't they just getting re-evaluated each time anyway?   Or is it  necessary for certain algorithms when clones are being put into special archives, or something like that?

**Comment by dhadka** [comment]
Objectives are copied because we often put solutions into different populations.  To keep the data independent, it creates copies.  For example, modifying one population won't modify the other population using the same solution since we created copies.

It explicitly doesn't copy attributes because you can put any arbitrary data there, which we may or may not be able to create copies.

Subclassing the solution would be better because you can override the copy() function.  However, you also need to make sure the problem's newSolution() method returns your subclass.

**Comment by fstonedahl** [comment]
Okay, thanks.  We'll consider the options.

**Comment by fstonedahl** [comment]
If we created a subclass of Solution called NoveltySolution that had a `double[] behavior` field and a `setBehavior()` method, then it seems to me that we won't be able to take advantage of the distributed computing of FutureSolution, right?

The `Problem.evaluate(...)` method will get handed a FutureSolution.  So if we attempt to cast it to NoveltySolution, that will cause an error... and it appears to me that there's no way to access the underlying inner Solution that a FutureSolution contains (presumably for good reason).

I may not have a good enough understanding how the parallel evaluation with FutureSolution works though... so perhaps I should just try it and see.

On the other hand, if we set a `behavior` **attribute** on a FutureSolution, I think that would work, because the SearchAlgorithm class is also seeing the FutureSolution objects (not the inner solutions), so it can access attributes that were set during the Problem.evaluate(...).

Thoughts?
